### PR TITLE
Revocable memory

### PR DESCRIPTION
**Pull request by pnowojski**
Add concept of revocable memory - memory that can be revoked when query exceeds memory limits. It was discussed in the context of the https://github.com/prestodb/presto/pull/5142 and https://github.com/prestodb/presto/issues/5144 as a long term solution. However it is independent pull request.

More work to do:
- rewrite partial aggregation to use revocable memory
- implement spillable hash aggregation to use revocable memory (instead of system memory)
- add "revoke memory" action triggered by coordinator when query exceeds `query.max-memory`
- add "revoke memory" action triggered by MemoryPool
- display revocable memory in UI

@cberner please take a look at first whether the fist commit is in the right direction.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 1]
```diff
@@ -158,4 +197,32 @@ public TaskContext addTaskContext(TaskStateMachine taskStateMachine, Session ses
         taskContexts.add(taskContext);
         return taskContext;
     }
+
+    private ListenableFuture<?> reserveMemoryUnsafe(long bytes)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: not synchronized, other private methods are

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 2]
```diff
@@ -158,4 +197,32 @@ public TaskContext addTaskContext(TaskStateMachine taskStateMachine, Session ses
...
+        if (willExceedMemoryLimit(bytes)) {
+            if (revocableFuture.isDone()) {
+                revocableFuture = SettableFuture.create();
+            }
+        }
+
+        ListenableFuture<?> future = memoryPool.reserve(queryId, bytes);
+
+        if (revocableFuture.isDone()) {
+            return future;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: is this branch required?

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 3]
```diff
@@ -158,4 +197,32 @@ public TaskContext addTaskContext(TaskStateMachine taskStateMachine, Session ses
...
+        }
+        else {
+            return Futures.allAsList(ImmutableList.of(revocableFuture, future));
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: what if there is no `revocableReserved` reserved? Then the returned future will never be done

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 4]
```diff
@@ -72,18 +83,33 @@ public synchronized void setResourceOvercommit()
...
+        ListenableFuture<?> future = reserveMemoryUnsafe(bytes);
         reserved += bytes;
         return future;
     }
 
+    public synchronized ListenableFuture<?> reserveRevocableMemory(long bytes)
+    {
+        checkArgument(bytes >= 0, "bytes is negative");
+
+        ListenableFuture<?> future = reserveMemoryUnsafe(bytes);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: so it is now possible to allocate above `maxMemory`? Previously exception was being thrown: https://github.com/prestodb/presto/pull/5711/commits/2a4d2197501b59e7c74ff3b863ff88483de7ab4e#diff-68ac3b293142f2c6f9d1f477192b9634L79.

I think we should wait for `revocableFuture` and make sure then we have enough free memory.

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 5]
```diff
@@ -97,7 +123,7 @@ public synchronized boolean tryReserveMemory(long bytes)
     {
         checkArgument(bytes >= 0, "bytes is negative");
 
-        if (reserved + bytes > maxMemory) {
+        if (willExceedMemoryLimit(bytes)) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: it uses different condition than in: https://github.com/prestodb/presto/pull/5711/commits/2a4d2197501b59e7c74ff3b863ff88483de7ab4e#diff-68ac3b293142f2c6f9d1f477192b9634L79

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java [hunk 6]
```diff
@@ -324,10 +325,20 @@ private void processNewSource(TaskSource source)
...
+            revokeMemoryIfNecessary();
             return processInternal();
         }
     }
 
+    private void revokeMemoryIfNecessary()
+    {
+        if (driverContext.isWaitingForRevocableMemory()) {
+            for (Operator operator : operators) {
+                operator.revokeMemory();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: I think you should check after each revoke that we are still waiting for revokable memory

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 7]
```diff
@@ -72,18 +83,33 @@ public synchronized void setResourceOvercommit()
...
+    public synchronized boolean isWaitingForRevocableMemory()
+    {
+        return !revocableFuture.isDone();
+    }
+
     public synchronized ListenableFuture<?> reserveMemory(long bytes)
     {
         checkArgument(bytes >= 0, "bytes is negative");
 
         if (reserved + bytes > maxMemory) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: extract and call it `willExceedNonRevokableMemoryLimit`

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 8]
```diff
@@ -158,4 +197,32 @@ public TaskContext addTaskContext(TaskStateMachine taskStateMachine, Session ses
         taskContexts.add(taskContext);
         return taskContext;
     }
+
+    private ListenableFuture<?> reserveMemoryUnsafe(long bytes)
+    {
+        if (willExceedMemoryLimit(bytes)) {
+            if (revocableFuture.isDone()) {
+                revocableFuture = SettableFuture.create();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: so allocating more revokable memory will cause revoke to current operators?

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java (hunk 1 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: this methods doesn't seem to be correct. Rethink it (think about different thread/operator scenarios).

File: presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java [hunk 9]
```diff
@@ -235,6 +248,17 @@ public void freeMemory(long bytes)
         memoryReservation.getAndAdd(-bytes);
     }
 
+    public void freeRevocableMemory(long bytes)
+    {
+        if (bytes == 0) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java]: why this special case?

File: presto-main/src/main/java/com/facebook/presto/operator/Operator.java [hunk 10]
```diff
@@ -73,6 +73,13 @@
     Page getOutput();
 
     /**
+     * After calling this method operator should revoke all reserved revocable memory
+     */
+    default void revokeMemory()
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/Operator.java]: shouldn't that return future too? Why block driver if operators might offload spill to background IO threads.

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java [hunk 11]
```diff
@@ -409,12 +397,48 @@ public OperatorStats getOperatorStats()
...
                 new Duration(finishUserNanos.get(), NANOSECONDS).convertToMostSuccinctTimeUnit(),
 
-                succinctBytes(memoryReservation.get()),
+                succinctBytes(memoryReservation.get() + revocableMemoryReservation.get()),
                 succinctBytes(systemMemoryContext.getReservedBytes()),
                 memoryFuture.get().isDone() ? Optional.empty() : Optional.of(WAITING_FOR_MEMORY),
                 info);
     }
 
+    private void handleOutOfMemory(ListenableFuture<?> future)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java]: please, provide better methods name :)

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java [hunk 12]
```diff
@@ -189,6 +189,23 @@ public long getTotalMemoryReservation()
     }
 
     @Override
+    public long getTotalRevocableMemoryReservation()
+    {
+        // acquire reference to outputStage before checking finalQueryInfo, because
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java]: what do revokable memory has with outputStage? Do we care about revokable memory reservation when the query finishes?

**Comment by kokosing** [comment]
Please provide a commit message that explains how revocable memory works. What is the expected behavior of allocation/deallocation with regards to memory configuration limits.

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 13]
```diff
@@ -158,4 +197,32 @@ public TaskContext addTaskContext(TaskStateMachine taskStateMachine, Session ses
         taskContexts.add(taskContext);
         return taskContext;
     }
+
+    private ListenableFuture<?> reserveMemoryUnsafe(long bytes)
+    {
+        if (willExceedMemoryLimit(bytes)) {
+            if (revocableFuture.isDone()) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: do you need `revocableFuture` at all? Even if the future is not done, you still ask `memoryPool` for memory (so there are still new memory requests even though we are in the middle of revoke). I think it might even cause deadlock in some scenarios.

Maybe you should ask `memoryPool` only when `revocableFuture` is done?

File: presto-main/src/main/java/com/facebook/presto/operator/Operator.java (hunk 10 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/Operator.java]: what will happen when operator `requestMemory` future is still pending? Is there a way to cancel such requests. If not, I think there might be a deadlock. Think about it.

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java (hunk 2 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: As discussed offline - yes

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java (hunk 3 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: In such case `revocableFuture` will always be done so this branch will never execute.

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java (hunk 4 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: As discussed offline - no, explained in commit message

File: presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java (hunk 9 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java]: I copied/pasted and modified `freeMemory` method. I think @yuananf has introduced this earlier this year.

File: presto-main/src/main/java/com/facebook/presto/operator/Operator.java (hunk 10 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/Operator.java]: I expect `revokeMemory` to work in background IO thread. There is no point in this method returning future completed when spilling has finished, since drivers/query will be blocked on out of memory future (which is released by freeing some memory). I expect background IO to call at some point `OperatorContext::freeRevocableMemory` or `OperatorContext::setRevocableMemoreReservation(0)`.

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java (hunk 11 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java]: Do you have any proposal? :)

**Comment by pnowojski** [comment]
@cberner when reviewing please take also a closer look at the second commit. I am not sure if there is some more elegant solution for revoking memory from `BLOCKED` drivers.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java (hunk 12 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java]: what does this comment refer to?

File: presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java [hunk 14]
```diff
@@ -236,6 +237,7 @@ public StageInfo getStageInfo(Supplier<Iterable<TaskInfo>> taskInfosSupplier, Su
 
             cumulativeMemory += taskStats.getCumulativeMemory();
             totalMemoryReservation += taskStats.getMemoryReservation().toBytes();
+            totalRevocableMemoryReservation += taskStats.getRevocableMemoryReservation().toBytes();
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java]: `totalMemoryReservation` name suggests that it includes the `revocable` memory. Consider renaming `totalMemoryReservation` to sth other.

File: presto-main/src/main/java/com/facebook/presto/operator/DriverStats.java [hunk 15]
```diff
@@ -196,6 +200,12 @@ public DataSize getMemoryReservation()
     }
 
     @JsonProperty
+    public DataSize getRevocableMemoryReservation()
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverStats.java]: is that visible in presto UI?

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java (hunk 12 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java]: to be honest I have no idea - I have copied paste `getTotalMemoryReservation` method

File: presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java (hunk 14 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java]: it's consistent with other places in code (like `StateStats`, `DriverStats`, etc) and with other stats. There is for example `totalCpuTime`, `totalBlockedTime` and `totalUserTime`. `total` means "sum of all values from child entities" (driver sums it's operators values, stage sums it's tasks etc...)

File: presto-main/src/main/java/com/facebook/presto/operator/DriverStats.java (hunk 15 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverStats.java]: not yet. It should be addressed somehow before merging.

**Comment by losipiuk** [comment]
@cberner 

We did some thinking on how revocable memory should work with global query memory limit `query.max-memory`.

Initially we thought that when coordinator notices that memory limit for a query is exceeded, but enough of that memory is revocable, it would trigger
revoking process on one or more worker nodes. This solution is tricky though as:
 * requires coordinator to react based on not up to date information
 * adds high coupling in execution between coordinator and worker. 
 * New problems arise: E.g. who is responsible to determine when revoking should be stopped. Worker nodes cannot do that reliably as they do not see state of whole cluster. It would require to add some extra execution state (maybe extend state space of `TaskStateMachine`?).
 * Generally speaking - it would add a lot of complexity in distributed job scheduling part of the system which is already non-trivial.

So we came with another approach. 

If memory revoking is enabled (there is a configuration and session switch to enable/disable revoking) we just do not count revokable memory as
used on coordinator node. We just assume that it will be handled by the worker nodes if `query.max-memory-per-node` limit is hit.
This is what is currently implemented in commit: b06b1806a702006c93a773a9c85b0058097fe060

With this approach query can use up to `query.max-memory-per-node`*`number_of_nodes` which is typically more than what we want. 
So we were thinking that if revoking is enabled, (as soon as both join and aggregation support spill to disk) we can set `query.max-memory-per-node` by default to
`query.max-memory` divided by number of nodes executing query. 
It would be sufficient to be sure that `query.max-memory` would not be exceeded for the query.

Though, that raises problem for queries operating on skewed data. 
These problems should be less visible with revoking, as queries would not fail but just execute longer due to more intensive disk usage, but still.

We thought this can be solved by changing `query.max-memory-per-node` adaptively while query is running. 
If coordinator notices situation:
 * that one node starts revoking memory
 * while others have plenty of memory free. 

It can lower the memory limit for those which do not need it - and increase it for one which uses all memory available.
While still keeping the sum of all `query.max-memory-per-node` equal to `query.max-memory`.

That is just a sketch - details to be decided - but we wanted to know if that is at least somewhat inline with what you were thinking resource management in presto should look like.

cc: @pnowojski

**Comment by cberner** [comment]
I think that mostly makes sense. I would suggest allocating the "revokable" memory from the system pool. That way it automatically will be excluded from the max memory calculations for the query, and you won't have to put any special cases in there. Since system memory is for performance, the workers can certainly make isolated decisions about when to start revoking.
I don't think trying to automatically set `query.max-memory-per-node` is a good idea, that sounds like a hack as spilling shouldn't have any direct impact on that setting

**Comment by losipiuk** [comment]
@cberner. 

TL;DR; can we talk on Monday? 8-11 P.S.T?

I am not sure if I like the fact to using system pool for revocable memory. 
It seemed to me that using system memory was intermediate step for spillable operators, until revocable memory is implemented.

If we used system memory how would we trigger the spill. 
Currently the `query.max-memory-per-node` applies only to user pool.
Should we add similar variable (or use same one) to mark the level of system-memory usage when spilling should start?

It could work but I think that users will have hard time on sizing user/system pool if we decide to use system pool for spillable operators. 
The cluster setup should be very different in case when queries are run spilling enabled vs spilling disabled.
When spilling is enabled we would need large system pool and small user pool.
And if spilling is disabled large user pool and small system pool.

Anyway. Let me try to explain our vision how it could work. And then let's try to map it to entities currently existing in codebase.

Ultimately we would rather see just one property for setting memory limit for query cluster-wise (`query.max-memory`).
And have presto automatically assign parts of this limit to individual nodes/operators participating in query. But let's discuss simpler 
solution for now.

Let's assume:
 * there is just one memory pool
 * query can borrow revocable and non-revocable memory from that pool
 * revocable and non-revocable memory is accounted separately 
There are two memory limits:
 * `query.max-memory`
 * `query.max-memory-per-node`

Ideally `query.max-memory` should be equal to `number_of_nodes_participating_in_query`*`query.max-memory-per-node`. But as we have data skew and `query.max-memory-per-node` is setup statically for now we can live with `number_of_nodes_participating_in_query`*`query.max-memory-per-node` > `query.max-memory`.

Spilling is triggering is managed locally by each node when memory used by query (revocable+non_revocable) exceeds `query.max-memory-per-node`.

Query is killed when memory used by query (only non_revocable) exceeds `query.max-memory`.

Eventually it would be nice to dynamically set the `query.max-memory-per-node` for individual nodes if one of nodes running query requires less memory and other requires more. You did not like this idea opting that spilling should have no impact on `query.max-memory-per-node`. But this in not spilling related at all. It is just general mechanism ensuring that system is more adaptable to runtime characteristics of query. This is totally separate work which may be pursued at its own pace. This moves us towards the aforementioned "ultimate solution" with only one memory-limit related config property.

Let's discuss how do current system and user pools and Facebook's vision play with this approach, and about approach itself.

cc: @pnowojski

**Comment by cberner** [comment]
Yep, I can do 10:15-11am Pacific Time on Monday.

I think triggering the spill would be pretty easy actually. We just set a threshold in the system pool, say 75%, at which it starts revoking memory. The reason I think the system pool is the right place for this is that the difference between the user pools (general & reserved) and the system pool is that system memory is used at the discretion of the engine to optimize performance. The buffers for spills are only for performance; we could disable them and always spill and still get correct results.

I don't think it's ever going to be possible to remove `query.max-memory-per-node` because that setting describes a fundamental limitation of the cluster (the amount of physical memory available on a single machine). Similarly I don't think making it dynamic makes sense, because the physical memory on the machine is not dynamic.

I think the place to make dynamic adjustments is in the sizing of the pools. If the general pool starts at 60% and the system pool at 40%, but the general pool usage never goes over 10%, then Presto should automatically change the ratio to something like 10%/90% at which point spill buffers could automatically become larger

**Comment by losipiuk** [comment]
> Yep, I can do 10:15-11am Pacific Time on Monday.

Cool let's meet then.

> I don't think it's ever going to be possible to remove `query.max-memory-per-node` because that setting describes a fundamental limitation of the cluster (the amount of physical memory available on a single machine). Similarly I don't think making it dynamic makes sense, because the physical memory on the machine is not dynamic.

Well - not really. `query.max-memory-per-node` defines how much memory can be consumed by one query. 
It is quite possible that from concurrently running queries different ones will utilize different memory amount.
Another case is when whole cluster is under-utilised at the moment and despite of conservatively setup `query.max-memory-per-node` limit - we could allow a query to consume more memory.
Cases like that allow for dynamic tweaking of the limit for individual query/node pairs.
Am I (we) missing something?

cc: @pnowojski

**Comment by pnowojski** [comment]
> I think triggering the spill would be pretty easy actually. We just set a threshold in the system pool, say 75%, at which it starts revoking memory. The reason I think the system pool is the right place for this is that the difference between the user pools (general & reserved) and the system pool is that system memory is used at the discretion of the engine to optimize performance. The buffers for spills are only for performance; we could disable them and always spill and still get correct results.

From the correctness perspective that would suffice. But think about performance. Once we implement spilling for both joins and aggregations, almost all queries will always finish (since almost all of the consumed memory will be revocable). The main concern will be performance. Now, without spilling and reasonable cluster configuration (memory limits adjusted to concurrency etc), as long as your queries are passing, they will run with optimal performance. If query exceeds its memory it fails. With spilling we need to control amount of the revocable memory assigned to queries, to make sure that some huge queries won't starve other possibly more important/priority/smaller queries.

Without limits on the revocable memory we are loosing control over queries. From a cluster admin perspective I would prefer setting some memory limit per query/user so that I can make sure, that for example some quick ad hoc queries executes as fast as possible without spilling while some other big query will not consume 90% of the memory on the cluster. I would be happy to eventually have some mechanism similar to CGroups tree hierarchy for queries/users.

> I don't think it's ever going to be possible to remove query.max-memory-per-node because that setting describes a fundamental limitation of the cluster (the amount of physical memory available on a single machine). Similarly I don't think making it dynamic makes sense, because the physical memory on the machine is not dynamic.

We don't want to remove this limit. But only to adjust it automatically. We came to such proposal when we tried to implement memory revoke mechanism on the coordinator. It's very complicated from the coordinator perspective because it always has outdated information about queries memory usage per nodes. Without revocable memory decision was simple - query exceeds `query.max-memeory-limit` it gets killed. With revocable memory it gets more complicated. Which node should we choose to revoke memory? Let's assume that we picked some worker(s). We issue them to start revoking the memory. When those worker should finish memory revoking? Should it revoke all memory? That would block whole query on the cluster for long time. Should it revoke only part of the memory? But what if in the meantime query memory usage on other nodes has increased? Coordinator should trigger another revoke on another worker? Maybe all of them? The only reasonable centralized solution that we could think of was to freeze whole query and revoke all of it's memory, which obviously is suboptimal.

That's why we started to think that decentralized solution for managing memory might be better. Every worker node controls the memory usage by itself and triggers memory revoking on its own, according to query/user priorities/limits. Worker has full, accurate and up to date knowledge about it's current state so it is much easier to control it on the worker then on the coordinator. Coordinator would only inform the worker about those priorities/limits. It makes both coordinator memory management simple (assign `query.max-memory-per-node` properties across the cluster and queries) and enforcing memory limits on the worker simple as well. In other words, worker would control that neither query exceeds `query.max-memory-per-node` limit by triggering revokes on some of the operator and coordinator would set dynamically `query.max-memory-per-node` properties to accommodate for potential data skew. 

It just seems to us as simpler, more elegant solution that also gives us much more control over memory allocation. Especially in the future. Another example of some fancy stuff that could be implemented in that model is that coordinator could deduce from blocked time/query cpu usage/http buffers, that on worker "A" operators are consuming lots of memory but are making little progress because query is blocked on some operators on different worker "B". Coordinator could then decrease `query.max-memory-per-node` limit on worker "A", forcing it to revoke this memory so that it can be used by other queries. Something like that is not even very difficult, since detecting bottlenecks by looking at http buffers should be easy.

**Comment by cberner** [comment]
In the case where the cluster is underutilized, I think the solution is to increase the value of `query.max-memory-per-node`. Rather than try to dynamically adjust that value, I think we should focus on automatically managing the workload that's running. For example, the new resource group system allows limits to be set on free memory after which new queries are no longer accepted (they become queued).
Also, in a cluster where spilling isn't enabled dynamically adjusting `query.max-memory-per-node` isn't an option since it would cause queries to fail, so I don't think it's a good idea to try to re-use that setting in the spilling code, by adjusting it to control buffer usage.

Workers can easily manage the system memory pool and implement a CGroups solution like you mentioned. They can get information both on the query priority and the resource group that it's running in. Trying to mix that into the general pool and `query.max-memory-per-node` isn't the right approach though as it mixes the concerns of memory management (allocating buffers for performance vs memory that's required for the execution of the query).
